controller/volume_controller.go

```diff
@@ -1026,6 +1031,114 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 	return nil
 }
 
+func (vc *VolumeController) ReconcileLocalReplica(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) error {
+	if v.Spec.DataLocality != types.DataLocalityBestEffort {
+		return nil
+	}
+
+	if v.Status.State != types.VolumeStateAttached {
+		return nil
+	}
+
+	if e == nil {
+		return fmt.Errorf("BUG: ReconcileLocalReplica needs a valid engine")
+	}
+
+	if !vc.isSafeToModifyReplica(v, e, rs) {
+		return nil
+	}
+
+	// hasLocalReplica indicate whether there exist a replica on the same node as engine
+	hasLocalReplica := false
+	for _, r := range rs {
+		if r.Spec.NodeID == e.Spec.NodeID {
+			hasLocalReplica = true
+		}
+	}
+
+	if !hasLocalReplica {
```

Comment:
call util func here, since doesn't seem that you depend on this variable anywhere else.

Reply:
Thank you. Fixed